File: image_matcher.py
import cv2
import numpy as np
import pandas as pd
from script_engine_utils import dist
from script_engine_utils import masked_mse

class ImageMatcher:
    def __init__(self):
        pass

    @staticmethod
    def template_match(detectObject,
                       screencap_im_bgr, screencap_search_bgr, screencap_mask_gray, screencap_outputmask_bgr, screencap_outputmask_gray,
                       detector_name, logs_path, script_mode, match_point,
                       output_cropping=None,
                       threshold=0.96, use_color=True, use_mask=True):
        # print(screencap_search.shape)
        if detector_name == "pixelDifference":
            matches, match_result, result_im_bgr = ImageMatcher.produce_template_matches(
                screencap_im_bgr.copy(),
                screencap_search_bgr.copy(),
                screencap_mask_gray,
                screencap_outputmask_bgr,
                logs_path,
                output_cropping=output_cropping,
                threshold=threshold,
                use_color=use_color,
                use_mask=use_mask
            )
        elif detector_name == "scaledPixelDifference":
            pass
            # matches, match_result, result_im_bgr = ImageMatcher.produce_scaled_template_matches(
            #     detectObject,
            #     screencap_im_bgr.copy(),
            #     screencap_search_bgr.copy(),
            #     screencap_outputmask_bgr,
            #     logs_path,
            #     threshold=threshold,
            #     use_color=use_color,
            #     use_mask=use_mask
            # )
        elif detector_name == "logisticClassifier":
            print("logistic detector unimplemented ")
            exit(0)
            # assets_folder = '/'.join((self.props['dir_path'] + '/' + action["actionData"]["positiveExamples"][0]["mask"]).split('/')[:-2])
            # matches,match_result,result_im = self.produce_logistic_matches(screencap_im.copy(), screencap_search, screencap_mask, logs_path, assets_folder)
        else:
            print("detector unimplemented! ")
            exit(0)

        h, w = screencap_outputmask_gray.shape[0:2]
        # plt.imshow(match_result, cmap='gray')
        # plt.title('Matching Result'), plt.xticks([]), plt.yticks([])
        cv2.imwrite(logs_path + 'matching_overlay.png', result_im_bgr)
        cv2.imwrite(logs_path + 'match_result.png', match_result * 255)
        # if self.props["scriptMode"] == "train":
        #     threshold = 0.94
        #     train_matches = []
        #     while not len(train_matches) > 0 and threshold > 0.1:
        #         threshold -= 0.02
        #         train_matches, train_match_result, train_result_im = self.produce_matches(screencap_im.copy(), screencap_search,
        #                                                                 screencap_mask, logs_path,
        #                                                                 threshold=threshold)
        #     # print(train_matches[0])
        #     print('train_threshold: ', threshold)
        #     cv2.imwrite(logs_path + 'matching_overlay_train.png', cv2.cvtColor(train_result_im, cv2.COLOR_BGR2RGB))
        #     cv2.imwrite(logs_path + 'match_result_train.png', train_match_result * 255)
        # print('pre sort matches: ', matches)
        matches.sort(reverse=True, key=lambda match: match[1])
        # print('post sort matches: ', matches)
        return [{
                'input_type': 'shape',
                'point': (match[0] + match_point[0], match[1] + match_point[1]) if match_point is not None else match,
                'shape': screencap_outputmask_gray,
                'matched_area': match_area,
                'height': h,
                'width': w,
                'score': score
        } for match, score, match_area in matches]


    @staticmethod
    def produce_template_matches(screencap_im_bgr, screencap_search_bgr, screencap_mask_gray, screencap_outputmask_bgr,
                                 logs_path, output_cropping=None, threshold=0.96, use_color=True, use_mask=True, script_mode='test'):
        # https://docs.opencv.org/3.4/de/da9/tutorial_template_matching.html
        h, w = screencap_search_bgr.shape[0:2]
        # print(screencap_search.shape)
        # print(screencap_mask.shape)
        # exit(0)
        # print(.shape)
        # exit(0)
        # print('match threshold: ', threshold)
        # exit(0)
        # cv2.imshow('screencap_im', screencap_im)
        # cv2.waitKey(0)
        # cv2.imshow('screencap_search', screencap_search)
        # cv2.waitKey(0)
        # cv2.imshow('screencap_mask', screencap_mask)
        # cv2.waitKey(0)
        # cv2.imshow('screencap_im', screencap_im_bgr)
        # cv2.waitKey(0)
        # cv2.imshow('screencap_search', screencap_search_bgr)
        # cv2.waitKey(0)
        try:
            match_result = cv2.matchTemplate(
                cv2.cvtColor(screencap_im_bgr.copy(), cv2.COLOR_BGR2GRAY) if not use_color else screencap_im_bgr,
                cv2.cvtColor(screencap_search_bgr.copy(), cv2.COLOR_BGR2GRAY) if not use_color else screencap_search_bgr,
                cv2.TM_CCOEFF_NORMED,result=None,
                mask=screencap_mask_gray if use_mask else None)
        except cv2.error as e:
            print(e)
            cv2.imshow('screencap_im', screencap_im_bgr)
            cv2.waitKey(0)
            cv2.imshow('screencap_search', screencap_search_bgr)
            cv2.waitKey(0)
            cv2.imshow('screencap_mask_gray', screencap_mask_gray)
            cv2.waitKey(0)
            print(screencap_im_bgr.shape, screencap_search_bgr.shape, screencap_mask_gray.shape)
            exit(1)
        # match_result = 1 - match_result

        # cv2.normalize(match_result, match_result, 0, 1, cv2.NORM_MINMAX, -1)
        # minVal, maxVal, minLoc, maxLoc = cv2.minMaxLoc(match_result, None)
        # match_result = cv2.matchTemplate(screencap_im, screencap_search, cv2.TM_CCOEFF_NORMED)
        dist_threshold = (w + h) * 0.1
        # print(dist_threshold)
        # print(np.where(match_result >= threshold))
        loc = np.where((np.inf > match_result) & (match_result >= threshold))
        # print('loc : ', loc)
        matches = []
        match_img_index = 1

        for pt in zip(*loc[::-1]):
            redundant = False
            match_score = match_result[pt[1], pt[0]]
            match_img_bgr = cv2.bitwise_and(screencap_im_bgr[pt[1]:pt[1] + h, pt[0]:pt[0] + w].copy(), screencap_outputmask_bgr)
            if output_cropping is not None:
                match_img_bgr = match_img_bgr[output_cropping[0][1]:output_cropping[1][1], output_cropping[0][0]:output_cropping[1][0]].copy()
            if match_score == np.inf:
                print(pt)
                continue
            for match_index in range(0, len(matches)):
                match = matches[match_index]
                match_coord = match[0]
                match_dist = dist(match_coord[0], match_coord[1], pt[0], pt[1])
                # print('dist ', match_dist)
                if match_dist < dist_threshold:
                    if match_score > match[1]:
                        matches[match_index] = (pt, match_score, match_img_bgr)
                    redundant = True
                    break
            if redundant:
                continue
            else:
                # print('{:f}'.format(match_result[pt[1], pt[0]]))
                matches.append((pt, match_score, match_img_bgr))
                if script_mode == "train":
                    # change name to fit image format
                    cv2.imwrite(logs_path + '-matched-' + str(match_img_index) + '-{:f}'.format(match_result[pt[1], pt[0]]) + '-img.png', match_img_bgr)
                match_img_index += 1
        best_match = str(np.max(match_result[np.where(np.inf > match_result)])) if (match_result[np.where(np.inf > match_result)]).size > 0 else 'none'
        print('n matches : ', len(matches), ' best match : ', best_match)
        with open(logs_path + '-best-' + best_match + '.txt', 'w') as log_file:
            log_file.write('n matches : ' + str(len(matches)))
        result_im_bgr = screencap_im_bgr
        for pt in zip(*loc[::-1]):
            cv2.rectangle(result_im_bgr, pt, (pt[0] + w, pt[1] + h), (0, 0, 255), 2)
        return matches, match_result, result_im_bgr

    # def produce_logistic_matches(self, screencap_im, screencap_search, screencap_mask, logs_path, assets_folder, threshold=0.7):
    #     # https://towardsdatascience.com/logistic-regression-using-python-sklearn-numpy-mnist-handwriting-recognition-matplotlib-a6b31e2b166a
    #     # https://stackoverflow.com/questions/5953373/how-to-split-image-into-multiple-pieces-in-python#:~:text=Copy%20the%20image%20you%20want%20to%20slice%20into,an%20image%20split%20with%20two%20lines%20of%20code
    #
    #     logistic_model = LogisticRegression()
    #     grayscale_screencap = cv2.cvtColor(screencap_im.copy(), cv2.COLOR_RGB2GRAY)
    #     # print(grayscale_screencap.shape)
    #     h,w = screencap_search.shape[:-1]
    #     imgs = []
    #     labels = []
    #     positive_examples = glob.glob(assets_folder + '/positiveExamples/*-img.png')
    #     negative_examples = glob.glob(assets_folder + '/negativeExamples/*-img.png')
    #     for positive_example in positive_examples:
    #         imgs.append(np.asarray(cv2.cvtColor(np.asarray(Image.open(positive_example)), cv2.COLOR_RGB2GRAY)).flatten())
    #         labels.append(1)
    #     for negative_example in negative_examples:
    #         imgs.append(np.asarray(cv2.cvtColor(np.asarray(Image.open(negative_example)), cv2.COLOR_RGB2GRAY)).flatten())
    #         labels.append(0)
    #     logistic_model.fit(imgs, labels)
    #     print(screencap_im.shape)
    #     print(h)
    #     print(w)
    #     # print(screencap_im[0:h, 0:w].shape)
    #     img_tiles = np.array([grayscale_screencap[y:y + h, x:x + w].flatten() for y in range(0, (grayscale_screencap.shape[0] - h)) for x in range(0, (grayscale_screencap.shape[1] - w))])
    #     print(img_tiles.shape)
    #     # logistic_model.predict(img_tiles).reshape(screencap_im.shape[0] - h, screencap_im.shape[1] - w)
    #
    #     print(screencap_search.shape)
    #     # print(screen_tiles)
    #     # for tile in enumerate(tiles):
    #     exit(0)
    #     return None,None,None

    @staticmethod
    def preprocess_scaled_match_object(state, detectObject):
        def createScaleFunction(scaleAction):
            return lambda x,y : scaleAction["actionData"]["xCoefficient"] * x + scaleAction["actionData"]["yCoefficient"] * y + scaleAction["actionData"]["constantTerm"]
        heightScaleAction = eval(detectObject["actionData"]["heightScale"], state.copy())
        heightScale = createScaleFunction(heightScaleAction)
        widthScaleAction = eval(detectObject["actionData"]["widthScale"], state.copy())
        widthScale = createScaleFunction(widthScaleAction)
        source_width = heightScaleAction["actionData"]["sourceWidth"]
        source_height = heightScaleAction["actionData"]["sourceHeight"]
        scaled_template_heights = np.zeros((source_height, source_width))
        scaled_template_widths = np.zeros((source_height, source_width))
        scaled_templates = np.empty((source_height, source_width), dtype=object)
        scaled_template_masks = np.empty((source_height, source_width), dtype=object)
        scaled_template_mask_sizes = np.zeros((source_height, source_width))
        scaled_template_output_masks = np.empty((source_height, source_width), dtype=object)
        scaled_template_output_masks_single_channel = np.empty((source_height, source_width), dtype=object)


        for y_index in range(0, source_height):
            for x_index in range(0, source_width):
                template_height = heightScale(x_index,y_index)
                template_width = widthScale(x_index,y_index)
                template_height = template_height if (y_index + template_height < source_height) else 0
                template_width = template_width if (x_index + template_width < source_width) else 0
                scaled_template_heights[y_index, x_index] = template_height
                scaled_template_widths[y_index, x_index] = template_width
                if template_width != 0 and template_height != 0:
                    scaled_templates[y_index, x_index] = cv2.resize(detectObject["actionData"]["positiveExamples"][0]["img"], (template_height, template_width), interpolation=cv2.INTER_LINEAR)
                    scaled_template_masks[y_index, x_index] = cv2.resize(detectObject["actionData"]["positiveExamples"][0]["mask"], (template_height, template_width), interpolation=cv2.INTER_LINEAR)
                    scaled_template_mask_sizes[y_index, x_index] = np.count_nonzero(detectObject["actionData"]["positiveExamples"][0]["mask_single_channel"]) * 3 * 255
                    scaled_template_output_masks = cv2.resize(detectObject["actionData"]["positiveExamples"][0]["outputMask"], (template_height, template_width), interpolation=cv2.INTER_LINEAR)
                    scaled_template_output_masks_single_channel = cv2.resize(detectObject["actionData"]["positiveExamples"][0]["outputMask_single_channel"], (template_height, template_width), interpolation=cv2.INTER_LINEAR)
                else:
                    scaled_templates[y_index, x_index] = None
                    scaled_template_masks[y_index, x_index] = None
                    scaled_template_mask_sizes[y_index, x_index] = 0
                    scaled_template_output_masks = None
                    scaled_template_output_masks_single_channel = None

        detectObject["actionData"]["scaled_templates"] = scaled_templates
        detectObject["actionData"]["scaled_template_heights"] = scaled_template_heights
        detectObject["actionData"]["scaled_template_widths"] = scaled_template_widths
        detectObject["actionData"]["scaled_template_masks"] = scaled_template_masks
        detectObject["actionData"]["scaled_template_mask_sizes"] = scaled_template_mask_sizes
        detectObject["actionData"]["scaled_template_output_masks"] = scaled_template_output_masks
        detectObject["actionData"]["scaled_template_output_masks_single_channel"] = scaled_template_output_masks_single_channel
        return detectObject




    @staticmethod
    def produce_scaled_template_matches(detectObject, screencap_im_bgr, screencap_search_bgr, screencap_outputmask_bgr,
                                 logs_path, threshold=0.96, use_color=True, use_mask=True, script_mode='test'):
        for y_index in range(0, detectObject["actionData"]["scaled_template_heights"].shape[0]):
            for x_index in range(0, detectObject["actionData"]["scaled_template_heights"].shape[1]):
                scaled_template = detectObject["actionData"]["scaled_templates"][y_index, x_index]
                if scaled_template is not None:
                    template_height = detectObject["actionData"]["scaled_template_heights"][y_index, x_index]
                    template_width = detectObject["actionData"]["scaled_template_widths"][y_index, x_index]
                    template_mask = detectObject["actionData"]["scaled_template_masks"][y_index, x_index]
                    template_mask_size = detectObject["actionData"]["scaled_template_mask_sizes"][y_index, x_index]
                    screencap_masked = cv2.bitwise_and(
                        screencap_im_bgr[y_index:y_index+template_height, x_index:x_index+template_width],
                        template_mask
                    )
                    ssim_coeff = masked_mse(screencap_masked, screencap_search_bgr, template_mask_size)

File: ImageMatcher/image_matcher.cpp
#include <opencv2/opencv.hpp>
#include <iostream>
#include <vector>
#include <cmath>
#include <map>
#include <string>
#include <tuple>
#include <algorithm>

const double MINIMUM_MATCH_PIXEL_SPACING = /*some_value*/;

#include <iostream>
#include <opencv2/opencv.hpp>
#include <map>
#include <string>
#include <vector>
#include <tuple>

std::vector<std::map<std::string, cv::Point>>
template_match(
    std::map<std::string, std::map<std::string, std::string>> detectObject,
    cv::Mat screencap_im_bgr,
    cv::Mat screencap_search_bgr,
    cv::Mat screencap_mask_gray,
    cv::Mat screencap_outputmask_bgr,
    cv::Mat screencap_outputmask_gray,
    std::string detector_name,
    std::string logs_path,
    std::string script_mode,
    cv::Point match_point,
    std::string log_level = "info",
    bool check_image_scale = true,
    cv::Rect output_cropping = cv::Rect(),
    double threshold = 0.96,
    bool use_color = true,
    bool use_mask = true) {

    std::vector<std::tuple<cv::Point, double, cv::Mat>> matches;
    cv::Mat match_result, result_im_bgr;

    if (detector_name == "pixelDifference") {
        std::tie(matches, match_result, result_im_bgr) = produce_template_matches(
            detectObject,
            screencap_im_bgr.clone(),
            screencap_search_bgr.clone(),
            screencap_mask_gray,
            screencap_outputmask_bgr,
            logs_path,
            std::stoi(detectObject["actionData"]["sourceScreenHeight"]),
            std::stoi(detectObject["actionData"]["sourceScreenWidth"]),
            log_level,
            check_image_scale,
            output_cropping,
            threshold,
            use_color,
            use_mask
        );
    }
    else if (detector_name == "scaledPixelDifference") {
        // Implement this case
    }
    else if (detector_name == "logisticClassifier") {
        std::cout << "logistic detector unimplemented" << std::endl;
        exit(0);
    }
    else {
        std::cout << "detector unimplemented!" << std::endl;
        exit(0);
    }

    int h = screencap_outputmask_gray.rows;
    int w = screencap_outputmask_gray.cols;

    if (log_level == "info") {
        cv::imwrite(logs_path + "matching_overlay.png", result_im_bgr);
        cv::imwrite(logs_path + "match_result.png", match_result * 255);
        cv::imwrite(logs_path + "output_mask.png", screencap_outputmask_gray);
    }

    int n_matches = matches.size();
    std::cout << "matches : ";

    for (const auto& [match, score, match_area] : matches) {
        std::cout << match << " ";
    }

    std::cout << match_point << std::endl;

    std::vector<std::map<std::string, cv::Point>> results;

    for (const auto& [match, score, match_area] : matches) {
        std::map<std::string, cv::Point> result;
        result["input_type"] = "shape";
        result["point"] = match_point == cv::Point() ? match : match + match_point;
        result["shape"] = screencap_outputmask_gray;
        result["matched_area"] = match_area;
        result["height"] = h;
        result["width"] = w;
        result["score"] = score;
        result["n_matches"] = n_matches;
        results.push_back(result);
    }

    return results;
}

std::tuple<std::vector<std::tuple<cv::Point, double, cv::Mat>>, cv::Mat, cv::Mat>
produce_template_matches(
    std::map<std::string, std::map<std::string, std::string>> detectObject,
    cv::Mat screencap_im_bgr,
    cv::Mat screencap_search_bgr,
    cv::Mat screencap_mask_gray,
    cv::Mat screencap_outputmask_bgr,
    std::string logs_path,
    int source_screen_height,
    int source_screen_width,
    bool check_image_scale,
    std::string log_level = "info",
    cv::Rect output_cropping = cv::Rect(),
    double threshold = 0.96,
    bool use_color = true,
    bool use_mask = true,
    std::string script_mode = "test") {

    int capture_height = screencap_im_bgr.rows;
    int capture_width = screencap_im_bgr.cols;
    bool is_dims_mismatch = check_image_scale && (capture_width != source_screen_width || capture_height != source_screen_height);
    double height_translation = 1.0;
    double width_translation = 1.0;
    cv::Mat match_result;
    cv::Mat mask;

    auto use_resized_im_only = detectObject["actionData"]["useImageRescaledToScreenOnly"] == "true";

    if (!use_resized_im_only) {
        try {
            cv::Mat image = use_color ? screencap_im_bgr : screencap_im_bgr.clone();
            cv::Mat templ = use_color ? screencap_search_bgr : screencap_search_bgr.clone();
            if (!use_color) {
                cv::cvtColor(image, image, cv::COLOR_BGR2GRAY);
                cv::cvtColor(templ, templ, cv::COLOR_BGR2GRAY);
            }
            mask = use_mask ? screencap_mask_gray : cv::Mat();
            cv::matchTemplate(image, templ, match_result, cv::TM_CCOEFF_NORMED, mask);
        } catch (cv::Exception& e) {
            std::cerr << "Error in matchTemplate: " << e.what() << std::endl;
            // handle error
        }
    }

    if (is_dims_mismatch || use_resized_im_only) {
        try {
            cv::Mat screencap_im_bgr_resized;
            cv::resize(screencap_im_bgr, screencap_im_bgr_resized, cv::Size(source_screen_width, source_screen_height), 0, 0, cv::INTER_AREA);

            cv::Mat image = use_color ? screencap_im_bgr_resized : screencap_im_bgr_resized.clone();
            cv::Mat templ = use_color ? screencap_search_bgr : screencap_search_bgr.clone();
            if (!use_color) {
                cv::cvtColor(image, image, cv::COLOR_BGR2GRAY);
                cv::cvtColor(templ, templ, cv::COLOR_BGR2GRAY);
            }
            mask = use_mask ? screencap_mask_gray : cv::Mat();
            cv::matchTemplate(image, templ, match_result, cv::TM_CCOEFF_NORMED, mask);

            height_translation = static_cast<double>(capture_height) / source_screen_height;
            width_translation = static_cast<double>(capture_width) / source_screen_width;

        } catch (cv::Exception& e) {
            std::cerr << "Error in matchTemplate: " << e.what() << std::endl;
            // handle error
        }
    }

    // Assuming filter_matches_and_get_result_im is implemented
    return filter_matches_and_get_result_im(
        detectObject,
        match_result,
        // ... other parameters ...
    );
}


double dist(double x1, double y1, double x2, double y2) {
    return std::sqrt((x2 - x1) * (x2 - x1) + (y2 - y1) * (y2 - y1));
}

std::pair<int, int> adjust_box_to_bounds(const cv::Point& pt, int box_width, int box_height, int screen_width, int screen_height, int box_thickness) {
    int x_overshoot = pt.x + box_width + box_thickness - screen_width;
    int y_overshoot = pt.y + box_height + box_thickness - screen_height;
    return {std::max(0, box_width - (x_overshoot > 0 ? x_overshoot : 0)),
            std::max(0, box_height - (y_overshoot > 0 ? y_overshoot : 0))};
}

std::tuple<std::vector<std::tuple<cv::Point, double, cv::Mat>>, cv::Mat, cv::Mat>
filter_matches_and_get_result_im(
    const cv::Mat& detectObject, const cv::Mat& match_result,
    std::vector<cv::Point> thresholded_match_results,
    const cv::Mat& screencap_im_bgr, const cv::Mat& screencap_search_bgr,
    const cv::Mat& screencap_outputmask_bgr,
    const std::string& logs_path, const std::string& log_level = "info",
    double height_translation = 1, double width_translation = 1,
    const std::string& script_mode = "test", cv::Rect output_cropping = cv::Rect(-1, -1, -1, -1)
) {
    int h = screencap_search_bgr.rows;
    int w = screencap_search_bgr.cols;

    double dist_threshold = std::max((w + h) * 0.1, MINIMUM_MATCH_PIXEL_SPACING);
    std::vector<std::tuple<cv::Point, double, cv::Mat>> matches;
    int match_img_index = 1;

    // Assuming match_result is a cv::Mat object and elements can be accessed by match_result.at<double>(row, col)
    for (const auto& pt : thresholded_match_results) {
        bool redundant = false;
        double match_score = match_result.at<double>(pt.y, pt.x);

        cv::Mat roi = screencap_im_bgr(cv::Rect(pt.x, pt.y, w, h)).clone();
        cv::Mat match_img_bgr;
        cv::bitwise_and(roi, screencap_outputmask_bgr, match_img_bgr);

        if (output_cropping.width > 0 && output_cropping.height > 0) {
            match_img_bgr = match_img_bgr(output_cropping).clone();
        }

        if (match_score == INFINITY) {
            continue;
        }

        double adjusted_pt_x = pt.x * width_translation;
        double adjusted_pt_y = pt.y * height_translation;

        for (size_t match_index = 0; match_index < matches.size(); ++match_index) {
            auto [match_coord, existing_match_score, _] = matches[match_index];
            double match_dist = dist(match_coord.x, match_coord.y, adjusted_pt_x, adjusted_pt_y);

            if (match_dist < dist_threshold) {
                if (match_score > existing_match_score) {
                    matches[match_index] = {cv::Point(adjusted_pt_x, adjusted_pt_y), match_score, match_img_bgr};
                }
                redundant = true;
                break;
            }
        }

        if (!redundant) {
            matches.push_back({cv::Point(adjusted_pt_x, adjusted_pt_y), match_score, match_img_bgr});
            match_img_index++;
        }
    }


    // Find best match
    std::string best_match = "none";
    cv::Point best_match_pt;
    cv::Mat inf_mask = (match_result != INFINITY);
    double minVal, maxVal;
    cv::Point minLoc, maxLoc;
    cv::minMaxLoc(match_result, &minVal, &maxVal, &minLoc, &maxLoc, inf_mask);
    
    if (maxVal != -INFINITY) {
        best_match_pt = maxLoc;
        best_match = std::to_string(maxVal);
    } else {
        std::cout << "a not valid" << std::endl;
    }

    std::cout << "n matches: " << matches.size() << " best match: " << best_match_pt << ", " << best_match << std::endl;

    cv::Mat result_im_bgr = screencap_im_bgr.clone();

    // Sort matches
    std::sort(matches.begin(), matches.end(),
              [](const auto& a, const auto& b) {
                  return std::get<1>(a) > std::get<1>(b);
              });

    // Drawing rectangles
    int box_w, box_h;
    for (const auto& pt : thresholded_match_results) {
        // Assuming adjust_box_to_bounds is a function that returns a pair of new dimensions (box_w, box_h)
        std::tie(box_w, box_h) = adjust_box_to_bounds(pt, w, h, screencap_im_bgr.cols, screencap_im_bgr.rows, 2);
        cv::rectangle(result_im_bgr, pt, cv::Point(pt.x + box_w, pt.y + box_h), cv::Scalar(0, 0, 255), 2);
    }

    if (thresholded_match_results.empty() && best_match_pt != cv::Point(-1, -1)) {
        std::tie(box_w, box_h) = adjust_box_to_bounds(best_match_pt, w, h, screencap_im_bgr.cols, screencap_im_bgr.rows, 2);
        cv::rectangle(result_im_bgr, best_match_pt, cv::Point(best_match_pt.x + box_w, best_match_pt.y + box_h), cv::Scalar(0, 0, 128), 2);
    }

    std::cout << "box_w, box_h: " << box_w << ", " << box_h << std::endl;

    return {matches, match_result, screencap_im_bgr};  // or whatever you need to return
}

int main() {
    // Your main function code here
    return 0;
}
